### PR TITLE
8270286: com.sun.net.httpserver.spi.HttpServerProvider: remove use of deprecated API

### PR DESCRIPTION
**Pull request by FrauBoes**
This is a cleanup fix that replaces the deprecated method `Class::newInstance()` with `cls.getDeclaredConstructor().newInstance()` when loading the provider.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8270286](https://bugs.openjdk.java.net/browse/JDK-8270286): com.sun.net.httpserver.spi.HttpServerProvider: remove use of deprecated API

### Reviewers
 * [Chris Hegarty](https://openjdk.java.net/census#chegar) (@ChrisHegarty - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4789/head:pull/4789` \
`$ git checkout pull/4789`

Update a local copy of the PR: \
`$ git checkout pull/4789` \
`$ git pull https://git.openjdk.java.net/jdk pull/4789/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4789`

View PR using the GUI difftool: \
`$ git pr show -t 4789`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4789.diff">https://git.openjdk.java.net/jdk/pull/4789.diff</a>

</details>

### CODE REVIEW

File: test/jdk/com/sun/net/httpserver/HttpServerProviderTest.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        @Override
+        public HttpsServer createHttpsServer(InetSocketAddress addr, int backlog) { return null; }
+    }
+
+    /**
+     * Test provider that is not public (NP)
+     */
+    static class ProviderNP extends HttpServerProvider {
+        // package-private default constructor
+        @Override
```
Comment (ChrisHegarty) [inline test/jdk/com/sun/net/httpserver/HttpServerProviderTest.java:132]: The class is pp here, while the default constructor ( as generated by the compiler  ) is public. So the comment needs a little adjustment, or better still just add an explicit no-args public constructor.

File: test/jdk/com/sun/net/httpserver/HttpServerProviderTest.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+ *      HttpServerProviderTest
+ * @run testng/othervm
+ *      -Dcom.sun.net.httpserver.HttpServerProvider=HttpServerProviderTest$ProviderPNPC
+ *      HttpServerProviderTest
+ * @run testng/othervm
+ *      -Dcom.sun.net.httpserver.HttpServerProvider=HttpServerProviderTest$ProviderNP
+ *      HttpServerProviderTest
+ * @run testng/othervm
+ *      -Dcom.sun.net.httpserver.HttpServerProvider=HttpServerProviderTest$ProviderT
+ *      HttpServerProviderTest
```
Comment (ChrisHegarty) [inline test/jdk/com/sun/net/httpserver/HttpServerProviderTest.java:39]: Maybe add a scenario where the given property is not a valid class, e..g DoesNotExist, or something.

**Review by ChrisHegarty — COMMENTED** [review]
The changes mostly look good. I have just a few small comments on the test.

File: test/jdk/com/sun/net/httpserver/HttpServerProviderTest.java (hunk 2 above)
Comment (FrauBoes) [reply test/jdk/com/sun/net/httpserver/HttpServerProviderTest.java:39]: Good point, thanks.

File: test/jdk/com/sun/net/httpserver/HttpServerProviderTest.java (hunk 1 above)
Comment (FrauBoes) [reply test/jdk/com/sun/net/httpserver/HttpServerProviderTest.java:132]: The default constructor has the same access modifier as the class (or is pp respectively), as per JLS 8.8.9. This being said, it's better to spell out the access control here, so I'll add explicit constructors to all test provider classes.

**Review by ChrisHegarty — APPROVED** [review]

**Pull request closed without merging**